### PR TITLE
fix(dyn-filter): filter left changes according to watermark before writing state table

### LINKED ISSUE #14014: make `WatermarkBufferByEpoch<const PERIOD: usize>`'s `PERIOD` as a virable
Currently, as the title, we have
https://github.com/risingwavelabs/risingwave/blob/420577906aa3ee39d4ba07589eb08a255a7a8156/src/stream/src/common/table/watermark.rs#L42

And it makes the constant period a part of the type `DefaultWatermarkBufferStrategy`.
It is especially annoying when we want to test the executor with no watermark buffer.

### LINKED ISSUE #17711: bug(dyn-filter): left side changes after state cleaning with right watermark
In DynamicFilter executor, we use watermarks on RHS to clean the state of left table, by `left_table.update_watermark(rhs_watermark)`. However in this way, later left changes will cause inconsistent table operations (e.g. double delete) on the left table, if the `update_watermark` did take effect.

An intuitive way to resolve this is to check whether the changed rows on left side is below watermark, and just ignore those rows that's below watermark. But another fact is that state cleaning watermark is not persisted, so after recovery we will temporarily have `None` watermark.

I guess for the sake of simplicity, we may just use state table with `inconsistent_op` for the left table of DynamicFilter.

(This is blocking #17694.)

### PR DESCRIPTION
**Pull request by stdrc**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

After #17767, we can get persisted watermark value for each vnode in state table. This PR utilizes this feature to filter out changes that are below committed table watermark before writing into state table. This fixes #17711.

close #14014 
## Checklist

- [x] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by st1page — APPROVED** [review]
LGTM

**Review by kwannoel — APPROVED** [review]
LGTM, thanks!

**Comment by stdrc** [comment]
* **#17816** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/17816?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @stdrc and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by stdrc** [comment]
Now this PR contains changes from #17885 and #17694, because they are related and a bit cyclic-dependent.

File: src/frontend/src/optimizer/plan_node/stream_dynamic_filter.rs [hunk 1]
```diff
@@ -100,7 +98,13 @@ impl StreamDynamicFilter {
                     // downstream. See `derive_watermark_columns`.
                     true
                 }
-                _ => false,
+                ExprType::LessThan | ExprType::LessThanOrEqual => {
+                    // For <= and <, watermark on rhs means all rows older than the watermark should already be emitted,
+                    // and future lhs inputs should be directly passed to downstream without any state table operation.
+                    // So, the state table can be cleaned up.
+                    true
+                }
```
Comment (stdrc) [inline src/frontend/src/optimizer/plan_node/stream_dynamic_filter.rs:106]: Now the left table is cleaned by rhs watermark when comparator is `<` and `<=` as well.

File: src/stream/src/executor/dynamic_filter.rs [hunk 2]
```diff
@@ -332,12 +351,23 @@ impl<S: StateStore, const USE_WATERMARK_CACHE: bool> DynamicFilterExecutor<S, US
...
+                        build_cond(self.comparator, committed_rhs_value.clone().flatten())
+                            .transpose()?;
+
+                    // The condition is `None` if there's no committed state cleaning watermark before.
+                    // Note that we should not use `state_cleaning_watermark` variable here, because
+                    // it represents outstanding watermark to be applied. Here we need the watermark
+                    // that has been applied to the state table, just like why we use `prev_epoch_value`
+                    // instead of `current_epoch_value` for `filter_condition`.
+                    let below_watermark_condition =
+                        build_cond(LessThan, self.left_table.get_committed_watermark().cloned())
```
Comment (stdrc) [inline src/stream/src/executor/dynamic_filter.rs:365]: Thanks to the removal of "state clean watermark buffer", now we can get up-to-date committed watermark from state table (and even after recovery).

**Pull request merged**